src/main/java/org/opensearch/ad/feature/ScriptMaker.java

```diff
@@ -0,0 +1,100 @@
+
+import org.opensearch.script.Script;
+import org.opensearch.script.ScriptType;
+
+public class ScriptMaker {
+    private static final String template = "\'%s\': doc['%s'].value";
+
+    /**
+     * We use composite aggregation for feature aggregation.  But composite aggregation
+     * does not support ordering results based on doc count, which is required by
+     * preview and historical related components.  We need to use terms aggregation.
+     * Terms aggregation does not support collecting terms from multiple fields in
+     * the same document.  Scripts come to the rescue: With a script to retrieve terms
+     * from multiple fields, we can still use terms aggregation to partition data.
+     * The script disables the global ordinals optimization and will be slower than
+     * collecting terms from a single field.  Still, it gives us the flexibility to
+     * implement this option at search time. For a simple example, consider the
+     * following query about the number field’s sum aggregation on buckets partitioned
+     * by category_field_1 and category_field_2 from index test.
+     *
+     * Query:
+     * GET /test/_search
+        {
+            "aggregations": {
+                "term_agg": {
+                    "terms": {
+                        "script": {
+                            "source": "['category_field_1': doc['category_field_1'].value,
+                             'category_field_2': doc['category_field_2'].value]",
+                            "lang": "painless"
+                        }
+                    },
+                    "aggregations": {
+                        "sum_number": {
+                            "sum": {
+                                "field": "number"
+                            }
+                        }
+                    }
+                }
+            }
+        }
+       *
+       * Result:
+       *"aggregations": {
+            "term_agg": {
+                "doc_count_error_upper_bound": 0,
+                "sum_other_doc_count": 0,
+                "buckets": [
+                    {
+                        "key": "{category_field_1=app_0, category_field_2=server_1}",
+                        "doc_count": 1,
+                        "sum_number": {
+                            "value": 1449.0
+                        }
+                    },
+                    {
+                        "key": "{category_field_1=app_1, category_field_2=server_1}",
+                        "doc_count": 1,
+                        "sum_number": {
+                            "value": 5200.0
+                        }
+                    },
+              ...
+       *
+       * I put two categorical field in a map for parsing the results.  Otherwise,
+       * I won't know which categorical value is for which field.
+     * @param fields categorical fields
+     * @return script to use in terms aggregation
+     */
+    public static Script makeTermsScript(List<String> fields) {
```

Comment:
Just confirm as I don't see how this class will be used in this PR. will we use this script for single category filed case? I think painless script query's performance is not so good.

Reply:
It is used here: https://github.com/kaituo/anomaly-detection-1/blob/multiCat4/src/main/java/org/opensearch/ad/feature/SearchFeatureDao.java#L202

Only use it for multi category field case.

Yeah, that's the only option if I continue to use terms aggregation. It is only used in Preview API.  So the performance impact is not wide.